server/tournament_service.py

```diff
@@ -0,0 +1,282 @@
+        tournament_game.state = TournamentGameState.RUNNING
+        self._games.remove(tournament_game)
+        await self.message_queue_service.publish(
+            config.MQ_EXCHANGE_NAME,
+            "tourneylauncher.createGame.success",
+            {
+                "game_id": game.id
+            },
+            correlation_id=tournament_game.request_id
+        )
+        metrics.matches_tournament.labels("success").inc()
+
+    async def _create_game(self, message: aio_pika.abc.AbstractIncomingMessage):
+        metrics.matches_tournament.labels("requested").inc()
+        try:
+            self._logger.info("Received Tournament game message")
+            await self._process_create_game(message)
+        except Exception:
+            self._logger.exception("Unknown failure creating tournament game")
+            await self.message_queue_service.publish(
+                config.MQ_EXCHANGE_NAME,
+                "tourneylauncher.createGame.failed",
+                {
+                    "request_id": message.correlation_id,
+                    "error_code": "OTHER"
+                },
+            )
+
+    async def _process_create_game(self, message):
+        body = message.body
+        body = json.loads(body)
+        game = TournamentGame(body)
+        assert game.request_id == message.correlation_id
+        if not await self._fetch_players(game):
+            await self._remove_and_cancel_game(game, make_idle=False)
+            return
+        _notify_players(game)
+
+    async def _fetch_players(self, game):
+        if not game.participants:
+            self._logger.warning("Tournament game requested with empty player list")
+            return False
+        await self.add_tournament_game(game)
+        for participant in game.participants:
+            player_id = participant["player_id"]
+            player = self.player_service[player_id]
+            if player is None:
+                self._logger.warning("Tournament game requested with player id(%s) that could not be found", player_id)
+                game.state = TournamentGameState.PLAYER_NOT_ONLINE
+                game.players_causing_cancel.add(player_id)
+                continue
+            if not player.state == PlayerState.IDLE:
+                self._logger.warning("Tournament game requested with player id(%s), player not idle", player_id)
+                game.state = TournamentGameState.PLAYER_NOT_IDLE
+                game.players_causing_cancel.add(player_id)
+            game.players.append(player)
+        return game.state == TournamentGameState.SCHEDULED
+
+    async def add_tournament_game(self, game):
+        self._games.add(game)
+
+    async def on_is_ready_response(self, message, player):
+        game: TournamentGame = await self._get_game_for_request_id(message["request_id"])
+        if game is None:
+            await player.lobby_connection.send_warning("You try to ready up for a game that does not exist")
```

Comment:
Generally these should be either `ClientError`'s or some custom message that can be localized by the client depending on whether or not reaching this code indicates a client bug.